File: validsudoku.cpp
//Time complexity : O(m * n)
//Space complexity : O(1)

//Approach:
            //Check each and every row and column first
            //If any of these are invalid, then return false
            //Check 9 squares, if anyone is invalid then return false
            //In the end return true

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        if(board.size() == 0)
            return true;
        
        //Check all the rows
        for(int i = 0 ; i < board.size(); i++)
        {
            bool checkArray[9] = {false};
            int j = 0;
            
            while(j < board[0].size())
            {
                if(isdigit(board[i][j]))
                {
                    int num = board[i][j] - '0';
                    if(checkArray[num - 1] == true)
                    {
                        return false;
                    }
                    checkArray[num - 1] = true;
                }
                j++;
            }
        }
        
        for(int i = 0 ; i < board.size(); i++)
        {
            bool checkArray[9] = {false};
            int j = 0;
            
            while(j < board[0].size())
            {
                if(isdigit(board[j][i]))
                {
                    int num = board[j][i] - '0';
                    if(checkArray[num - 1] == true)
                    {
                        return false;
                    }
                    checkArray[num - 1] = true;
                }
                j++;
            }
        }
        
        //check for 3*3 matrices
        vector<vector<int>> dirs = {{0 , 1}, {0 , -1}, {1 , 0}, {-1 , 0}, {1 , 1}, {-1 , 1}, {1, -1}, {-1 , -1}};
        vector<vector<int>> centerCells = {{1 , 1}, {1 , 4}, {1 , 7}, {4, 1}, {4, 4}, {4, 7}, {7 , 1}, {7 , 4}, {7 , 7}};
        
        for(auto center : centerCells)
        {
            bool checkArray[9] = {false};
            int r = center[0];
            int c = center[1];
            if(isdigit(board[r][c]))
            {
                int num = board[r][c] - '0';
                if(checkArray[num - 1] == true)
                {
                    return false;
                }
                checkArray[num - 1] = true;
            }
            
            for(auto dir : dirs)
            {
                int nr = dir[0] + r;
                int nc = dir[1] + c;
                if(isdigit(board[nr][nc]))
                {
                    int num = board[nr][nc] - '0';
                    if(checkArray[num - 1] == true)
                    {
                        return false;
                    }
                    checkArray[num - 1] = true;
                } 
            }
        }
        
        return true;
    }
};

File: ValidSudoku.cs
using System;
namespace Algorithms
{
    /// Time Complexity :  O(N2)
    // Space Complexity :O(N) 
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No
    public class ValidSudoku
    {
        public bool IsValidSudoku(char[][] board)
        {
            if (board == null || board.Length == 0) return false;
            int N = 9;

            int[,] rows = new int[N, N];
            int[,] cols = new int[N, N];
            int[,] boxes = new int[N, N];

            for (int r = 0; r < N; r++)
            {
                for (int c = 0; c < N; c++)
                {
                    if (board[r][c] == '.')
                    {
                        continue;
                    }
                    int pos = board[r][c] - '1';

                    // Check the row
                    if (rows[r, pos] == 1)
                    {
                        return false;
                    }
                    rows[r, pos] = 1;

                    // Check the column
                    if (cols[c, pos] == 1)
                    {
                        return false;
                    }
                    cols[c, pos] = 1;

                    // Check the box
                    int idx = (r / 3) * 3 + c / 3;
                    if (boxes[idx, pos] == 1)
                    {
                        return false;
                    }
                    boxes[idx, pos] = 1;
                }
            }
            return true;



        }
    }
}
